x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/ExpressionBuilder.java

```diff
@@ -729,21 +730,27 @@ public Literal visitDecimalLiteral(DecimalLiteralContext ctx) {
     public Literal visitIntegerLiteral(IntegerLiteralContext ctx) {
         Tuple<Source, String> tuple = withMinus(ctx);
 
-        long value;
+        Number value;
         try {
-            value = Long.valueOf(StringUtils.parseLong(tuple.v2()));
+            value = StringUtils.parseIntegral(tuple.v2());
         } catch (QlIllegalArgumentException siae) {
             throw new ParsingException(tuple.v1(), siae.getMessage());
         }
 
-        Object val = Long.valueOf(value);
-        DataType type = DataTypes.LONG;
-        // try to downsize to int if possible (since that's the most common type)
-        if ((int) value == value) {
-            type = DataTypes.INTEGER;
-            val = Integer.valueOf((int) value);
+        DataType type;
+        if (value instanceof BigInteger) {
+            type = DataTypes.UNSIGNED_LONG;
+        } else {
+            assert value instanceof Long : "Expected value [" + value + "] of type Long but got: " + value.getClass();
```

Comment:
Assert should probably be replaced with an Exception?

Reply:
Refactored away.